fe1-web/model/network/method/message/Message.ts

```diff
@@ -121,42 +121,52 @@ export class Message {
    * @param witnessSignatures- The signatures of the witnesses
    * @returns - The created message
    */
-  public static fromData(
+  public static async fromData(
     data: MessageData, witnessSignatures?: WitnessSignature[],
   ): Promise<Message> {
     const encodedDataJson: Base64UrlData = encodeMessageData(data);
     let publicKey = KeyPairStore.getPublicKey();
     let privateKey = KeyPairStore.getPrivateKey();
+    let signature: Signature;
 
-    if (isSignedWithToken(data)) {
-      return getCurrentPopTokenFromStore().then((token) => {
-        if (token) {
-          publicKey = token.publicKey;
-          privateKey = token.privateKey;
-        } else {
-          console.error('Impossible to sign the message with a pop token: no token found for '
-            + 'current user in this LAO');
-        }
-        const signature: Signature = privateKey.sign(encodedDataJson);
-
-        return new Message({
-          data: encodedDataJson,
-          sender: publicKey,
-          signature,
-          message_id: Hash.fromStringArray(encodedDataJson.toString(), signature.toString()),
-          witness_signatures: (witnessSignatures === undefined) ? [] : witnessSignatures,
-        });
+    const messageObjectMap = isMessageSignedWithToken.get(data.object);
+    if (!messageObjectMap) {
+      throw new Error(`Message signature for object ${data.object} is unsupported.`);
+    }
+
+    const isSignedWithToken = messageObjectMap.get(data.action);
```

Comment:
Couldn't you make your life easier by doing `x.get(data.object)?.get(data.action)` and skipping one level of error ?

Reply:
True, but I'll use an interface anyway since this solution is preferable.